### PR TITLE
T232 React components TS conversion

### PR DESCRIPTION
**Pull request by mmahdigh**

### CODE REVIEW

File: src/providers/toast.tsx [hunk 1]
```diff
@@ -73,9 +98,10 @@ function ToastProvider({ children }) {
           <li key={message.id}>
             <Toast
               delay={TOAST_DELAY}
+              // @ts-ignore
               variant={message.variant}
```
Comment (mmahdigh) [inline src/providers/toast.tsx:102]: Per `Toast` component types (and my quick glance at the docs), A `variant` prop doesn't seem to be available. However there's a `bg` prop that does probably what we have in mind. 
But it's possible that a `variant` prop is available under the hood and it's just not documented. Just wanted to double-check.

File: src/client/arweave/sync/diff-merge-logic.ts [hunk 2]
```diff
@@ -152,45 +169,46 @@ const mergeSpecialTags = (acc, batch) => {
...
+ * @param oldMetadata
+ * @param newMetadata
+ * @param primaryKey Primary key to survive the diff; see hasMetadata().
  *   TODO: pending T244, change to 'id'.
- * @returns {Object} The newMetadata omitting each { prop: value } already present in oldMetadata
+ * @returns The newMetadata omitting each { prop: value } already present in oldMetadata
  *   and empty props are ignored.
  */
-export function rightDiff(oldMetadata = {}, newMetadata = {}, primaryKey = 'subscribeUrl') {
+export function rightDiff<T extends Episode | Podcast>(oldMetadata : T,
```
Comment (mmahdigh) [inline src/client/arweave/sync/diff-merge-logic.ts:204]: `rightDiff` accepts both episode and podcast metadata if I'm not mistaken?

If so, I suggest that we break it down into two separate functions (one accepting episodes, and one podcasts). At the current state, types are unnecessarily complicated and hard to get right.

File: src/utils.ts [hunk 3]
```diff
@@ -60,31 +67,33 @@ export function toDate(date) {
...
 
 /**
- * @param {Object} metadata
- * @returns {boolean} true if `metadata` has specific metadata other than:
+ * @param metadata
+ * @returns true if `metadata` has specific metadata other than:
  *   `subscribeUrl`, `publishedAt` and an empty episodes list
  */
-export function hasMetadata(metadata) {
+export function hasMetadata(metadata: Partial<Podcast> | Partial<Episode>) {
```
Comment (mmahdigh) [inline src/utils.ts:74]: Like `rightDiff`, `hasMetadata` also seems to accept a wide variety of inputs which makes it pretty hard to type it properly.

File: src/client/arweave/sync/diff-merge-logic.ts [hunk 4]
```diff
@@ -61,41 +72,44 @@ function mergeEpisodesMetadata(oldEpisodes, newEpisodes) {
...
 /**
- * @param {Array.<Array.<Object>>} episodeBatches
- * @returns {Array.<Object>}
+ * @param episodeBatches
+ * @returns
  */
-export function mergeEpisodeBatches(episodeBatches) {
-  return episodeBatches.reduce((mergedEps, batch) => mergeEpisodesMetadata(mergedEps, batch), {});
+export function mergeEpisodeBatches(episodeBatches: PartialEpisodeWithDate[][]) {
+  return episodeBatches.reduce((mergedEps, batch) => mergeEpisodesMetadata(mergedEps, batch), []);
```
Comment (mmahdigh) [inline src/client/arweave/sync/diff-merge-logic.ts:97]: I also change `{}` to `[]` in here. `{}`as the initial value didn't make sense to me since `mergeEpisodesMetadata` also needed arrays of metadata as its parameters.

File: src/client/arweave/sync/diff-merge-logic.ts [hunk 5]
```diff
@@ -61,41 +72,44 @@ function mergeEpisodesMetadata(oldEpisodes, newEpisodes) {
...
-      oldEpisodesWithMerges[oldEpisodeIndex] = mergeEpisodeMetadata(oldEpisode, newEpisode);
+      oldEpisodesWithMerges[oldEpisodeIndex] = mergeEpisodeMetadata(
+        oldEpisode, newEpisode) as PartialEpisodeWithDate;
     }
   }
   return newEpisodes
     .filter((_, index) => !duplicateNewEpisodeIndices.includes(index))
     .concat(oldEpisodesWithMerges)
-    .sort((a, b) => b.publishedAt - a.publishedAt);
+    .sort((a, b) => b.publishedAt.getTime() - a.publishedAt.getTime());
```
Comment (mmahdigh) [inline src/client/arweave/sync/diff-merge-logic.ts:89]: `b.publishedAt - a.publishedAt` is correct in JS but TS requires that we either call `getTime` method (like I did) or use [unary plus ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Unary_plus)to explicitly convert it to a number. (`+b.publishedAt - +a.publishedAt`)

File: src/client/arweave/sync/diff-merge-logic.ts (hunk 2 above)
Comment (decentdav) [reply src/client/arweave/sync/diff-merge-logic.ts:205]: Indeed, throughout my code, I've been dancing around not having typed variables. Let's add some more functions!

File: src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+      return !!value;
+  }
+}
+
+/**
+ * 
+ * @param obj is an object that might be empty/undefined
+ * @returns true if the given array or object is not empty
+ */
+export function isNotEmpty<T extends object>(obj: T | undefined | null | {}) : obj is T {
```
Comment (mmahdigh) [inline src/utils.ts:165]: I essentially tried to create a [User-Defined Type Guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) for scenarios like `getPodcastFeed` function.

Obviously it'd have been better if I could do this while keeping the functionality unaffected but below code didn't work:

```
export function isEmpty<T extends object>(obj: T | {}) obj is {} {
  return !obj || typeof obj !== 'object' || Object.keys(obj).length === 0;
}
```
But the compiler doesn't understand consider `not being {}` with `being T` equal for some reason.

**Review by decentdav — APPROVED** [review]
Nice work!

File: src/components/pod-graph/cytoscape/graph-logic/interfaces/interfaces.ts [hunk 7]
```diff
@@ -1,20 +1,33 @@
-export interface Podcast {
+export interface Podcast extends PodcastTags, PodcastContent {}
```
[the remark this answers is not in the document]
Comment (mmahdigh) [reply src/components/pod-graph/cytoscape/graph-logic/interfaces/interfaces.ts:1]: That's good idea. Will fix it in the next patch.

File: src/utils.ts (hunk 6 above)
Comment (mmahdigh) [reply src/utils.ts:165]: Then `obj is T` would be wrong. If `obj` is empty then technically it can't be of type T. This would be correct:

```

Comment (decentdav) [reply src/utils.ts]: Alright, let's just go with the `isNotEmpty()` for now.

File: src/providers/arweave.tsx [hunk 8]
```diff
@@ -8,20 +8,38 @@ import { SubscriptionsContext } from './subscriptions';
 import * as arweave from '../client/arweave';
 import * as arsync from '../client/arweave/sync';
 import { isNotEmpty, valuesEqual, concatMessages } from '../utils';
+import { JWKInterface } from 'arweave/node/lib/wallet';
+import Transaction from 'arweave/node/lib/transaction';
+import { TransactionStatus } from '../client/arweave/sync';
+
+interface ArweaveContaxtType {
```
Comment (decentdav) [inline src/providers/arweave.tsx:15]: You have a small ContaxtTypo

File: src/providers/arweave.tsx [hunk 9]
```diff
@@ -158,13 +177,13 @@ function ArweaveProvider({ children }) {
         walletAddress,
         loadNewWallet,
         prepareSync,
-        hasPendingTxs,
+        hasPendingTxs: hasPendingTxs(),
```
Comment (decentdav) [inline src/providers/arweave.tsx:180]: Why is this change needed? Does this not run the risk of returning an outdated value for `hasPendingTxs()`?

File: src/providers/arweave.tsx (hunk 8 above)
Comment (mmahdigh) [reply src/providers/arweave.tsx:15]: Fixed.

File: src/providers/arweave.tsx (hunk 9 above)
Comment (mmahdigh) [reply src/providers/arweave.tsx:180]: No the result won't be stale. It just makes more sense that `hasPendingTxs` be a boolean rather than a function.

File: src/client/interfaces.ts [hunk 10]
```diff
@@ -24,15 +24,22 @@ export interface PodcastSeed {
   tags: PodcastTags,
 }
 
+export interface PodcastDTO extends Omit<Podcast, 'firstEpisodeDate' | 'lastEpisodeDate'
```
Comment (decentdav) [inline src/client/interfaces.ts:27]: Nice, I will start incorporating more utility types as well.

**Review by decentdav — APPROVED** [review]
Good work!
